src/azul/chalice.py

```diff
@@ -145,3 +158,163 @@ def _log_response(self, response):
     # Some type annotations to help with auto-complete
     lambda_context: LambdaContext
     current_request: Request
+
+
+class ValidatingAzulChaliceApp(AzulChaliceApp):
+
+    def __init__(self, app_name, unit_test=False, spec=None):
+        super().__init__(app_name=app_name, unit_test=unit_test, spec=spec)
+        self.route_specs: MutableMapping[str, RouteSpec] = {}
+
+    def route(self, *args, validate: bool = True, **kwargs):
+        """
+        Decorates a view handler function in a Chalice application. Refer to
+        docstring from super class for additional information.
+
+        :param validate: If False, do not validate requests submitted to the endpoint.
+        """
+
+        azul_chalice_decorator = super().route(*args, **kwargs)
+
+        def decorator(view_func):
+            # noinspection PyShadowingNames
+            @functools.wraps(view_func)
+            def f(*args, **kwargs):
+                if validate:
+                    self._validate()
+                return view_func(*args, **kwargs)
+
+            return azul_chalice_decorator(f)
+
+        return decorator
+
+    def _validate(self):
+        path = self.current_request.context['resourcePath']
+        self.route_specs[path].validate_params(self.current_request)
+
+    def _register_spec(self,
+                       path: str,
+                       path_spec: Optional[JSON],
+                       method_spec: Optional[JSON],
+                       methods: Iterable[str]):
+        """
+        Add a route's specifications to the specification object.
+        """
+        if path not in self.route_specs:
+            assert path not in self.specs['paths'], 'Only specify path_spec once per route path'
+            route_spec = RouteSpec(path, path_spec)
+            self.route_specs[path] = route_spec
+        else:
+            route_spec = self.route_specs[path]
+        if method_spec is not None:
+            for method in methods:
+                # OpenAPI requires HTTP method names be lower case
+                method = method.lower()
+                route_spec.set_route_spec(method, method_spec)
+        if len(route_spec.methods) > 0:
+            self.specs['paths'][path] = route_spec.to_dict()
+
+
+class RouteSpec:
+    """
+    Assists with managing details about API routes, provides access to request
+    validation.
+    """
+
+    def __init__(self, path, path_spec=None):
+        self.path: str = path
+        self.specs: MutableMapping[str, Dict[Any, Any]] = {}
+        self.validators: MutableMapping[str, Mapping[str, SpecValidator]] = {}
+
+        if path_spec:
+            self.path_spec = copy_json(path_spec)
+        else:
+            self.path_spec = {}
+
+    @property
+    def methods(self):
+        return [method for method in self.specs.keys()]
```

Comment:
One problem with generating a list here is that a new list will be allocated in memory each time this property is accessed. A simpler solution would just be
```suggestion
        return self.specs.keys()
```
which only returns a reference to the same object.

This new method not only saves memory space, but it's faster too. The property is used to check membership which is O(n) on a list, but O(1) on a set or dict (or dict view object in this case).

Reply:
That's a good catch, ill try to use this more often.